src/lightning/pytorch/callbacks/model_checkpoint.py

```diff
@@ -514,19 +514,23 @@ def _format_checkpoint_name(
 
         # check and parse user passed keys in the string
         groups = re.findall(r"(\{.*?)[:\}]", filename)
-        if len(groups) >= 0:
-            for group in groups:
-                name = group[1:]
 
-                if auto_insert_metric_name:
-                    filename = filename.replace(group, name + "={" + name)
+        # sort keys from longest to shortest to avoid replacing substring
+        # eg: if keys are "epoch" and "epoch_test", the latter must be replaced first
+        groups = sorted(groups, key=lambda x: len(x), reverse=True)
```

Comment:
Thanks for this comment. Now I finally get what root problem with the reported bug is 😅